### PR TITLE
Stateful Detection of Black-Box Adversarial Attacks

### LINKED ISSUE #1373: Stateful Detection of Black-Box Adversarial Attacks
**Algorithm described here**
https://arxiv.org/pdf/1907.05587.pdf

**Implementation**
As a part of engineering thesis I'm implementing this algorithm. In addition, I will provide example usages of new module including: jupyter notebook and kubeflow pipeline.

### PR DESCRIPTION
**Pull request by twrdyyy**
# Description
Close #1373 
@beat-buesser 
As proposed in https://arxiv.org/pdf/1907.05587.pdf paper.
This PR introduces new defence agains attacks on black-box classifier.  Example notebook of usage is provided inside the PR in **notebooks/defence_black_box_pytorch_example.ipynb**

New classes introduced:
1. **BlackBoxDetector** - performs scan on given query
 -> art.defences.detector.evasion.black_box.detector 
2. **MemoryQueue** - Used by detector, saves encoded query, has fixed length and is implemented fully in numpy library
 -> art.defences.detector.evasion.black_box.memory_queue
3. **NearestNeighborsWrapper** - Wrapper for sklearn Nearest Neighbor class to be properly used in ART
 -> art.defences.detector.evasion.black_box.knn_wrapper

There are also new loss function named *contrastive_loss*, implemented purely in numpy, in tensorflow and in pytorch.

## Type of change

Please check all relevant options.

- [X] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [X] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

**I do not know if we need some unit tests for this implementation, if we do please comment out so I will add them!**

# Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by twrdyyy** [comment]
@beat-buesser The DCO CI wanted me to sign-off the commits

**Comment by twrdyyy** [comment]
Please review the changes and guide me what to do to properly integrate it to 1.9.0 branch 😄

**Comment by beat-buesser** [comment]
Thank you very much for your pull request, we will review it as soon as possible.

**Comment by beat-buesser** [comment]
Hi @twrdyyy It looks like you have accidentally signed commits of other authors. Could you please update your branch to only have your own signed commits on top of `dev_1.9.0`?

**Comment by twrdyyy** [comment]
Hey, Yes I will correct this, how about the code?

**Comment by twrdyyy** [comment]
@beat-buesser I've rebased this branch to dev_1.9.0 version and have updated all the commits

**Comment by beat-buesser** [comment]
Before we take a closer look at the code, could you please update the notebook to reproduce the data point for Boundary Attack in Table 1? Please add as many comments to the notebook as possible to explain each step.

**Comment by twrdyyy** [comment]
I will update the notebook today, but what exactly do you mean by Table 1? And by adding Boundary Attack you mean to include this after the HopSkipJump attack sure? @beat-buesser

**Comment by beat-buesser** [comment]
With Table 1 I meant the results of [Chen et al.](https://arxiv.org/pdf/1907.05587.pdf ) presented in their Table 1 and discussed in their section 4.2. I think it would be good to see if the if this pull request can achieve a comparable number of detections (e.g. their 288).

**Comment by twrdyyy** [comment]
@beat-buesser I've added Boundary attack to notebook and have done some cleanup. Both encoder and classifier models were added to /notebooks/models as weights to be loaded inside example notebook. If you would like to see the comparison with table 1, I did some tests. For Boundary attack for over 1000 attack on MNIST data:
* num of queries - 54752
* attacks detected - 1000
* detection success rate - 100%

For threshold = 0.003, K-neighbours = 20 and max_queue size = 100

I did not include this in notebook as it takes some time to calculate.

File: art/defences/detector/evasion/black_box/contrastive_loss.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+Contrastive loss function as proposed in paper:
+| Paper link: https://arxiv.org/abs/1907.05587
```
Comment (beat-buesser) [inline art/defences/detector/evasion/black_box/contrastive_loss.py:20]:
```suggestion
Contrastive loss functions

| Paper link: https://arxiv.org/abs/1907.05587
```

File: art/defences/detector/evasion/black_box/contrastive_loss.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+"""
+import numpy as np
+import torch
+import tensorflow as tf
+from math import sqrt
```
Comment (beat-buesser) [inline art/defences/detector/evasion/black_box/contrastive_loss.py:25]:
```suggestion
from math import sqrt

import numpy as np
import tensorflow as tf
import torch
```

File: art/defences/detector/evasion/black_box/contrastive_loss.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+def np_contrastive_loss(
+    e_i: np.ndarray,
+    e_p: np.ndarray,
+    e_j: np.ndarray,
+    e_n: np.ndarray,
+    m: float = sqrt(10),
+) -> np.ndarray:
+    """
+    Contrastive loss function implemented for numpy as proposed in paper:
```
Comment (beat-buesser) [inline art/defences/detector/evasion/black_box/contrastive_loss.py:36]: Change format to ART docstring style and add params and returns.

File: art/defences/detector/evasion/black_box/contrastive_loss.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+def torch_contrastive_loss(
+        e_i: torch.Tensor,
+        e_p: torch.Tensor,
+        e_j: torch.Tensor,
+        e_n: torch.Tensor,
+        m: float = sqrt(10),
+) -> torch.Tensor:
+    """
+    Contrastive loss function implemented for pytorch as proposed in paper:
```
Comment (beat-buesser) [inline art/defences/detector/evasion/black_box/contrastive_loss.py:57]: Change format to ART docstring style and add params and returns.

File: art/defences/detector/evasion/black_box/contrastive_loss.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+def tf_contrastive_loss(
+        e_i: tf.Tensor,
+        e_p: tf.Tensor,
+        e_j: tf.Tensor,
+        e_n: tf.Tensor,
+        m: float = sqrt(10),
+) -> tf.Tensor:
+    """
+    Contrastive loss function implemented for tensorflow as proposed in paper:
```
Comment (beat-buesser) [inline art/defences/detector/evasion/black_box/contrastive_loss.py:78]: Change format to ART docstring style and add params and returns.

File: art/defences/detector/evasion/black_box/contrastive_loss.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+Contrastive loss function as proposed in paper:
+| Paper link: https://arxiv.org/abs/1907.05587
+"""
+import numpy as np
+import torch
+import tensorflow as tf
```
Comment (beat-buesser) [inline art/defences/detector/evasion/black_box/contrastive_loss.py:24]: Move imports of TensorFlow and PyTorch into the functions requiring them to make the module framework independent. Use `if TYPE_CHECKING:` for typing.

File: art/defences/detector/evasion/black_box/detector.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+Stateful detection on Black-Box (2021).
+| Paper link: https://arxiv.org/abs/1907.05587
```
Comment (beat-buesser) [inline art/defences/detector/evasion/black_box/detector.py:20]:
```suggestion
Stateful detection on Black-Box  Adversarial Attacks.

| Paper link: https://arxiv.org/abs/1907.05587
```

File: art/defences/detector/evasion/black_box/detector.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+from art.estimators.estimator import NeuralNetworkMixin, BaseEstimator
+from art.utils import CLIP_VALUES_TYPE
+
+logger = logging.getLogger(__name__)
+
+
+class BlackBoxDetector(NeuralNetworkMixin, BaseEstimator, ABC):
+    """
+    Stateful detection on Black-Box by Twardy F. (2021).
+    | Paper link: https://arxiv.org/abs/1907.05587
```
Comment (beat-buesser) [inline art/defences/detector/evasion/black_box/detector.py:42]:
```suggestion
    Stateful detection on Black-Box Adversarial Attacks.
    
    | Paper link: https://arxiv.org/abs/1907.05587
```

File: art/defences/detector/evasion/black_box/detector.py [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+    def __init__(
+        self,
+        similarity_encoder: Union[PyTorchEncoder, TensorFlowEncoder],
+        memory_queue: MemoryQueue,
+        knn: NearestNeighborsWrapper,
+        detection_threshold: float,
+    ):
+        """
+        Stateful detection on Black-Box (2021).
+        | Paper link: https://arxiv.org/abs/1907.05587
```
Comment (beat-buesser) [inline art/defences/detector/evasion/black_box/detector.py:54]:
```suggestion
        Stateful detection on Black-Box Adversarial Attacks.
        
        | Paper link: https://arxiv.org/abs/1907.05587
```

File: art/estimators/encoding/pytorch.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
```
Comment (beat-buesser) [inline art/estimators/encoding/pytorch.py:3]:
```suggestion
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2021
```

File: art/estimators/encoding/pytorch.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class PyTorchEncoder(EncoderMixin, PyTorchEstimator):
+    """
+    This class implements an encoder model using the PyTorch framework.
+    """
+
+    def __init__(
```
Comment (beat-buesser) [inline art/estimators/encoding/pytorch.py:45]: Apply Black formatter with line length 120.

File: art/defences/detector/evasion/black_box/memory_queue.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+"""
+import logging
+import numpy as np
+from typing import Tuple
+
+
+logger = logging.getLogger(__name__)
+
+
+class MemoryQueue:
```
Comment (beat-buesser) [inline art/defences/detector/evasion/black_box/memory_queue.py:29]: Could you please add an abstract base class for `Memory` to enable different implementations?

File: art/defences/detector/evasion/black_box/memory_queue.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+        :param max_size: Maximum number of elements stored in memory queue
+        :param initial_elements: Initial elements for memory queue
+        """
+        self._max_size = max_size
+        self._memory = initial_elements
+        self._element_shape = element_shape
+
+    def extend(self, elements: np.ndarray):
+        """
+        Extends queue with new elements
```
Comment (beat-buesser) [inline art/defences/detector/evasion/black_box/memory_queue.py:52]: Please add params to docstring.

File: art/defences/detector/evasion/black_box/memory_queue.py [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    def __init__(
+        self,
+        max_size: int,
+        element_shape: Tuple[int, ...],
+        initial_elements: np.ndarray = None,
+    ):
+        """
+        Memory queue for black-box detector.
+
+        :param max_size: Maximum number of elements stored in memory queue
```
Comment (beat-buesser) [inline art/defences/detector/evasion/black_box/memory_queue.py:43]: Add `element_shape`.

File: art/defences/detector/evasion/black_box/knn_wrapper.py [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+
+class NearestNeighborsWrapper:
+    """
+    Wrapper to sklearn NearestNeighbors class
+    """
+
+    def __init__(
+        self,
+        k_neighbors: int,
+        memory_queue: MemoryQueue,
```
Comment (beat-buesser) [inline art/defences/detector/evasion/black_box/knn_wrapper.py:41]: Change `MemoryQueue` to new abstract base class.

File: art/defences/detector/evasion/black_box/knn_wrapper.py [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements wrapper for sklearn knn algorithm.
+"""
+import logging
+from typing import Callable, Union, Tuple, Any
+
+import numpy as np
+from sklearn.neighbors import NearestNeighbors
```
Comment (beat-buesser) [inline art/defences/detector/evasion/black_box/knn_wrapper.py:25]: Move the import of sklearn inside of the class/methods requiring it to make the module framework-independent.

File: art/defences/detector/evasion/black_box/knn_wrapper.py [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+        self._k_neighbors = k_neighbors
+        self._distance_function = distance_function
+        self._memory_queue = memory_queue
+        self._knn = NearestNeighbors(n_neighbors=self._k_neighbors)
+
+    def __call__(
+        self, query: np.ndarray, with_distances: bool = False
+    ) -> Union[Tuple[np.ndarray, Any], np.ndarray]:
+        """
+        :param query: Query for which we calculate mean distance
```
Comment (beat-buesser) [inline art/defences/detector/evasion/black_box/knn_wrapper.py:60]: Add params.

File: art/defences/detector/evasion/black_box/knn_wrapper.py [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+        :returns: Mean distance between K nearest neighbors and query
+        """
+        k_distances, _ = self._knn.kneighbors(query)
+        k_distance = np.mean(k_distances)
+        if with_distances:
+            return k_distance, k_distances
+        return k_distance
+
+    def update_memory(self):
```
Comment (beat-buesser) [inline art/defences/detector/evasion/black_box/knn_wrapper.py:69]: Add docstring.

File: art/defences/detector/evasion/black_box/knn_wrapper.py [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+import numpy as np
+from sklearn.neighbors import NearestNeighbors
+
+from art.defences.detector.evasion.black_box.contrastive_loss import np_contrastive_loss
+from art.defences.detector.evasion.black_box.memory_queue import MemoryQueue
+
+logger = logging.getLogger(__name__)
+
+
+class NearestNeighborsWrapper:
```
Comment (beat-buesser) [inline art/defences/detector/evasion/black_box/knn_wrapper.py:33]: To enable different nearest neighbor searches:
- Move module to `art.estimators.neighbors.scikit.learn.py`
- Create abstract mixin-class `NeighborMixin` defining abstract method `predict` to replace the current `__call__` and `fit to replace the current `update_memory`.
- Update method names in this class

Comment (beat-buesser) [inline art/defences/detector/evasion/black_box/knn_wrapper.py:33]:
```suggestion
class ScikitlearnNearestNeighbors:
```

File: art/defences/detector/evasion/black_box/knn_wrapper.py [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements wrapper for sklearn knn algorithm.
```
Comment (beat-buesser) [inline art/defences/detector/evasion/black_box/knn_wrapper.py:19]:
```suggestion
This module implements estimator for scikit-learn NearestNeighbors.
```

File: art/defences/detector/evasion/black_box/detector.py [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+from art.defences.detector.evasion.black_box.memory_queue import MemoryQueue
+from art.estimators.encoding import TensorFlowEncoder
+from art.estimators.encoding.pytorch import PyTorchEncoder
+from art.estimators.estimator import NeuralNetworkMixin, BaseEstimator
+from art.utils import CLIP_VALUES_TYPE
+
+logger = logging.getLogger(__name__)
+
+
+class BlackBoxDetector(NeuralNetworkMixin, BaseEstimator, ABC):
```
Comment (beat-buesser) [inline art/defences/detector/evasion/black_box/detector.py:39]:
```suggestion
class StatefulBlackBoxDetector(BaseEstimator):
```

File: art/defences/detector/evasion/black_box/__init__.py [hunk 22]
```diff
@@ -0,0 +1,11 @@
...
+This module implements Stateful detection on black-box adversarial attacks
+"""
+from art.defences.detector.evasion.black_box.detector import BlackBoxDetector
+from art.defences.detector.evasion.black_box.memory_queue import MemoryQueue
+from art.defences.detector.evasion.black_box.knn_wrapper import NearestNeighborsWrapper
+from art.defences.detector.evasion.black_box.contrastive_loss import (
+    torch_contrastive_loss,
+    np_contrastive_loss,
+    tf_contrastive_loss
+)
```
Comment (beat-buesser) [inline art/defences/detector/evasion/black_box/__init__.py:11]:
```suggestion
```

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @twrdyyy 
Please take a look at my suggestions above, some minor changes and relocating classes for future extensions.

After these changes, please create a test module for each new module in `tests` using **pytest**.

Please also take a look at the log of the Style Check CI run and fix their alerts if needed.

The notebook look much better now, maybe see if you can write a few sentences around the most important cells to describe what is happening and add links to the paper, etc. Shortly before merging we can move the models to our large file storage for the notebook support data.

**Comment by twrdyyy** [comment]
@beat-buesser I will make those changes during the weekend, because I had to finish my paper during last two weeks. Thanks for review! I will make everything as you said.

**Comment by twrdyyy** [comment]
@beat-buesser I am sorry for being inactive here, but I had some major problems with paper and finishing the thesis I see that ART is updated to 1.9 so I will do the following changes to 1.10

**Comment by beat-buesser** [comment]
Hi @twrdyyy Sounds good, let me know when I can help.

**Comment by beat-buesser** [comment]
Please reopen pull request if work continues.

**Pull request closed without merging**